File: ios/Sources/CapWeb3AuthPlugin/CapWeb3AuthPlugin.swift
import Foundation
import Capacitor
import Web3Auth

/**
 * Please read the Capacitor iOS Plugin Development Guide
 * here: https://capacitorjs.com/docs/plugins/ios
 */
@objc(CapWeb3AuthPlugin)
public class CapWeb3AuthPlugin: CAPPlugin, CAPBridgedPlugin {
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "echo", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "login", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "logout", returnType: CAPPluginReturnPromise),
    ]

    public let identifier = "CapWeb3AuthPlugin"
    public let jsName = "CapWeb3Auth"
    private let implementation = CapWeb3Auth()

    @objc func echo(_ call: CAPPluginCall) {
        let value = call.getString("value") ?? ""
        call.resolve([
            "value": implementation.echo(value)
        ])
    }

    private var web3auth: Web3Auth? = nil
    
    @objc func logout(_ call: CAPPluginCall) {
        Task {
            do {
                try await web3auth?.logout();
            } catch {
                call.reject("Error logging out through Web3", nil)
            }
        } 
        call.resolve()
    }

    @objc func login(_ call: CAPPluginCall) {
        guard let clientId = call.getString("clientId") else {
            call.reject("clientId is required")
            return
        }

        guard let network = call.getString("network") else {
            call.reject("network is required")
            return
        }

        guard let provider = call.getString("provider") else {
            call.reject("provider is required")
            return
        }

        let loginHint = call.getString("loginHint")
        let redirectUrl = call.getString("redirectUrl")

        var w3aNetwork: Network
        switch network.lowercased() {
        case "sapphire_devnet":
            w3aNetwork = .sapphire_devnet
        default:
            w3aNetwork = .sapphire_mainnet
        }
        
        Task {
            do {
                // Initialize Web3Auth if not already initialized
                if web3auth == nil {
                    print("Initializing Web3Auth...")
                    web3auth = try! await Web3Auth(W3AInitParams(
                        clientId: clientId,
                        network: w3aNetwork,
                        redirectUrl: redirectUrl ?? ""
                    ))
                    print("Web3Auth initialized successfully")
                }
                
                guard let web3auth = web3auth else {
                    call.reject("Failed to initialize Web3Auth")
                    return
                }
                
                var loginParams: W3ALoginParams
                
                switch provider.lowercased() {
                case "google":
                    print("Attempting Google login")
                    loginParams = W3ALoginParams(loginProvider: .GOOGLE)
                case "apple":
                    print("Attempting Apple login")
                    loginParams = W3ALoginParams(loginProvider: .APPLE)
                case "x":
                    print("Attempting Apple login")
                    loginParams = W3ALoginParams(loginProvider: .TWITTER, redirectUrl: redirectUrl ?? "")
                case "email":
                    if let email = loginHint {
                        print("Attempting Email login with: \(email)")
                        loginParams = W3ALoginParams(
                            loginProvider: .EMAIL_PASSWORDLESS,
                            extraLoginOptions: .init(login_hint: email)
                        )
                    } else {
                        call.reject("loginHint is required for email login")
                        return
                    }
                default:
                    call.reject("Unsupported login provider: \(provider)")
                    return
                }
                
                print("Executing login with params...")
                let result = try await web3auth.login(loginParams)
                print("Login successful")
                
                // Get user info
                let userInfo = try web3auth.getUserInfo()
                let privateKey = web3auth.getPrivkey()
                
                let userInfoDict: [String: String] = [
                    "aggregateVerifier": userInfo.aggregateVerifier ?? "",
                    "email": userInfo.email ?? "",
                    "name": userInfo.name ?? "",
                    "profileImage": userInfo.profileImage ?? "",
                    "typeOfLogin": userInfo.typeOfLogin ?? "",
                    "verifier": userInfo.verifier ?? "",
                    "verifierId": userInfo.verifierId ?? "",
                    "dappShare": userInfo.dappShare ?? "",
                    "idToken": userInfo.idToken ?? "",
                    "oAuthIdToken": userInfo.oAuthIdToken ?? "",
                    "oAuthAccessToken": userInfo.oAuthAccessToken ?? ""
                ]
                
                let resultData: [String: Any] = [
                    "ed25519PrivKey": "",
                    "privKey": privateKey ?? "",
                    "sessionId": "",
                    "userInfo": userInfoDict
                ]
                
                call.resolve(["result": resultData])
                
            } catch {
                print("Login failed with error: \(error.localizedDescription)")
                call.reject("Login failed: \(error.localizedDescription)")
            }
        }
    }

}

File: android/src/main/java/com/kingcyk/capweb3auth/CapWeb3AuthPlugin.kt
package com.kingcyk.capweb3auth

import android.content.Intent
import android.net.Uri
import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin
import com.web3auth.core.Web3Auth
import com.web3auth.core.types.ExtraLoginOptions
import com.web3auth.core.types.LoginParams
import com.web3auth.core.types.Network
import com.web3auth.core.types.Provider
import com.web3auth.core.types.Web3AuthOptions
import kotlinx.coroutines.future.await
import kotlinx.coroutines.*

@CapacitorPlugin(name = "CapWeb3Auth")
class CapWeb3AuthPlugin : Plugin() {
    private val implementation = CapWeb3Auth()
    private var web3auth: Web3Auth? = null

    @PluginMethod
    fun echo(call: PluginCall) {
        val value = call.getString("value")

        val ret = JSObject()
        ret.put("value", implementation.echo(value))
        call.resolve(ret)
    }

    override fun handleOnNewIntent(intent: Intent?) {
        super.handleOnNewIntent(intent)
        web3auth?.setResultUrl(intent?.data)
    }

    @PluginMethod
    fun logout(call: PluginCall) {
        web3auth?.logout()
        call.resolve()
    }

    @PluginMethod
    fun login(call: PluginCall) {
        val clientId = call.getString("clientId")
        val network = call.getString("network")

        if (clientId == null) {
            call.reject("clientId is required")
            return
        }

        if (network == null) {
            call.reject("network is required")
            return
        }

        var w3aNetwork: Network? = null
        when (network) {
            "sapphire_devnet" -> w3aNetwork = Network.SAPPHIRE_DEVNET
            else -> {
                w3aNetwork = Network.SAPPHIRE_MAINNET
            }
        }

        val provider = call.getString("provider")
        val loginHint = call.getString("loginHint")
        val redirectUrl = call.getString("redirectUrl")

        val loginParams = when (provider) {
            "google" -> LoginParams(Provider.GOOGLE)
            "apple" -> LoginParams(Provider.APPLE)
            "x" -> LoginParams(Provider.TWITTER)
            "email" -> LoginParams(Provider.EMAIL_PASSWORDLESS, extraLoginOptions = ExtraLoginOptions(
                login_hint = loginHint
            ))
            else -> {
                call.reject("Unknown provider")
                return
            }
        }

        if (web3auth == null) {
            web3auth = Web3Auth(
                Web3AuthOptions(
                    clientId = clientId,
                    network = w3aNetwork,
                    redirectUrl = Uri.parse(redirectUrl)
                ),
                context = this@CapWeb3AuthPlugin.context
            )
            web3auth?.setResultUrl(this@CapWeb3AuthPlugin.activity.intent?.data)
            val initCompletableFuture = web3auth?.initialize()
            initCompletableFuture?.whenComplete { _, error ->
                if (error == null) {
                    println("web3auth init success")
                    val loginCompletableFuture = web3auth?.login(loginParams)
                    loginCompletableFuture?.whenComplete { result, error ->
                        if (error == null) {
                            val ret = JSObject()
                            val resultObj = JSObject()
                            val userInfo = JSObject()
                            userInfo.put("email", result?.userInfo?.email ?: "")
                            userInfo.put("name", result?.userInfo?.name ?: "")
                            userInfo.put("profileImage", result?.userInfo?.profileImage ?: "")
                            userInfo.put("verifier", result?.userInfo?.verifier ?: "")
                            userInfo.put("verifierId", result?.userInfo?.verifierId ?: "")
                            userInfo.put("typeOfLogin", result?.userInfo?.typeOfLogin ?: "")
                            userInfo.put("aggregateVerifier", result?.userInfo?.aggregateVerifier ?: "")
                            userInfo.put("dappShare", result?.userInfo?.dappShare ?: "")
                            userInfo.put("idToken", result?.userInfo?.idToken ?: "")
                            userInfo.put("oAuthIdToken", result?.userInfo?.oAuthIdToken ?: "")
                            userInfo.put("oAuthAccessToken", result?.userInfo?.oAuthAccessToken ?: "")
                            resultObj.put("userInfo", userInfo)
                            resultObj.put("privKey", result?.privKey ?: "")
                            resultObj.put("ed25519PrivKey", result?.ed25519PrivKey ?: "")
                            resultObj.put("sessionId", result?.sessionId ?: "")
                            ret.put("result", resultObj)
                            call.resolve(ret)
                        } else {
                            println("web3auth login fail")
                            call.reject(error.message)
                        }
                    }
                } else {
                    println("web3auth init fail")
                    call.reject(error.message)
                }
            }
        } else {
            val loginCompletableFuture = web3auth?.login(loginParams)
            loginCompletableFuture?.whenComplete { result, error ->
                if (error == null) {
                    println("web3auth login success")
                    val ret = JSObject()
                    val resultObj = JSObject()
                    val userInfo = JSObject()
                    userInfo.put("email", result?.userInfo?.email ?: "")
                    userInfo.put("name", result?.userInfo?.name ?: "")
                    userInfo.put("profileImage", result?.userInfo?.profileImage ?: "")
                    userInfo.put("verifier", result?.userInfo?.verifier ?: "")
                    userInfo.put("verifierId", result?.userInfo?.verifierId ?: "")
                    userInfo.put("typeOfLogin", result?.userInfo?.typeOfLogin ?: "")
                    userInfo.put("aggregateVerifier", result?.userInfo?.aggregateVerifier ?: "")
                    userInfo.put("dappShare", result?.userInfo?.dappShare ?: "")
                    userInfo.put("idToken", result?.userInfo?.idToken ?: "")
                    userInfo.put("oAuthIdToken", result?.userInfo?.oAuthIdToken ?: "")
                    userInfo.put("oAuthAccessToken", result?.userInfo?.oAuthAccessToken ?: "")
                    resultObj.put("userInfo", userInfo)
                    resultObj.put("privKey", result?.privKey ?: "")
                    resultObj.put("ed25519PrivKey", result?.ed25519PrivKey ?: "")
                    resultObj.put("sessionId", result?.sessionId ?: "")
                    ret.put("result", resultObj)
                    call.resolve(ret)
                } else {
                    println("web3auth login fail")
                    call.reject(error.message)
                }
            }
        }
    }
}
